CONTRIBUTING.md

```diff
@@ -576,6 +575,60 @@ that are part of this project but not production code. The canonical example
 of this is `junit`.</dd>
 </dl>
 
+
+Reviewing and accepting your contribution
+-----------------------------------------
+
+We review every contribution carefully to ensure that the change is of high
+quality and fits well with the rest of the Elasticsearch codebase. If accepted,
+we will merge your change and usually take care of backporting it to
+appropriate branches ourselves.
+
+We really appreciate everyone who is interested in contributing to
+Elasticsearch and regret that we sometimes have to reject contributions even
+when they might appear to make genuine improvements to the system. Reviewing
+contributions can be a very time-consuming task, yet the review team is small
+and our time is very limited. In some cases the time we would need to spend on
+reviews would outweigh the benefits of a change by a significant margin.
+
+We strongly recommend you discuss your change in a Github issue before spending
+much time on its implementation. We may have to reject your contribution if it
+takes the wrong approach to solving a problem or if it solves a problem that
+does not need solving. We do not enjoy seeing time wasted like that.
```

Comment:
I'm on the fence about this suggestion from an anonymous third party:

```suggestion
does not need solving.
```

It's really bad if someone spends ages on a bad change and I want to communicate some sympathy here since it does happen sometimes, but maybe there's a better way.

Reply:
Maybe something more long these lines?

```suggestion
does not need solving. We don't want anyone to feel like they have
wasted their time.
```

Reply:
Thanks @pugnascotia, I reworded this paragraph quite a bit and this sentence is now roughly the one that says `An up-front discussion would have saved a good deal of wasted time in these cases.`